File: 501.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.maxCount = 0
        self.count = 0
        self.lastVal = None
        self.res = list()

        def inorder(node):
            if node is None:
                return
            inorder(node.left)
            if node.val != self.lastVal:
                self.count = 1
                self.lastVal = node.val
            else:
                self.count += 1
            if self.count == self.maxCount:
                self.res.append(self.lastVal)
            elif self.count > self.maxCount:
                self.maxCount = self.count
                self.res[:] = [self.lastVal]
            inorder(node.right)

        inorder(root)
        return self.res

File: sol/501/501.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct Solution {
    void inorder(TreeNode *node, int &lastVal, int &count, int &maxCount, vector<int> &res){
        if (node == nullptr){
            return;
        }
        inorder(node->left, lastVal, count, maxCount, res);
        if (node->val != lastVal){
            count = 1;
            lastVal = node->val;
        } else {
            ++count;
        }
        if (count > maxCount){
            maxCount = count;
            res.clear();
        }
        if (count == maxCount){
            res.push_back(lastVal);
        }
        inorder(node->right, lastVal, count, maxCount, res);
    }

    vector<int> findMode(TreeNode* root) {
        vector<int> res;
        int lastVal = -100001, count = 0, maxCount = 0;
        inorder(root, lastVal, count, maxCount, res);
        return res;
    }
};
